### PR TITLE
FlatBuffers post

### PR DESCRIPTION
**Pull request by alexpusch**

### CODE REVIEW

File: flatbuffers/flatbuffers_case_study.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+Case study of Javascript Flat Buffers
+================================
+
+Data serialization is very common process in web application. So common that no one thinks of it. You just pull data from the DB, and your server and client frameworks just handle this thing for you. The server serializes the data into JSON, and the client parses it to JS object. This flow is simple and natural but, as everything, when taken to the extreme start to show scaling issues.
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md]: the first sentence reads weird..
maybe:
Data serialization is a very common process in web applications

Comment (BlueHotDog) [reply flatbuffers/flatbuffers_case_study.md]: I think a lot of the benefit are related to data de-serialization as well.

File: flatbuffers/flatbuffers_case_study.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+Whats an 'extreme' for data serialization? A lot of data. Serializing and deseriazling tens of thousands of objects takes it toll from the server, and more importantly, from the single threaded browser page.
+
+A possible solution for this scaling issue are FlatBuffers. We've utilized this tool to do our heavy lifting and lived to tell about it.
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md]: is Flatbuffers? it's a single technology

File: flatbuffers/flatbuffers_case_study.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+* Recordings storage
+* Recordings player
+
+The recorder have the ability to capture any event or DOM mutation in a page, and stream the recorded entities to our servers. As you can imagine a session might consist of a huge number of events. Every mouse move, every scroll, every change to the page itself must be recorded. This can easily mount to ten or hundreds of thousands of data points.
+
+All this data must be stored and accessible for the recorders counterpart - the player. The player knows how to take the mass of events and reconstruct them into a watchable clip.
+
+This undertaking raises many challenges, one of which is to transfer all of those events from the recorded user to our servers, and from our servers to the player. The efficiency of this process is measured both by the size of the transfered data and by the time it takes to parse it, measures that JSON does not excel in.
+
+# What are FlatBuffers?
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md:27]: Flatbuffer vs other solutions?

File: flatbuffers/flatbuffers_case_study.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
+Case study of Javascript Flat Buffers
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md]: seems this is more of a flatbuffer vs json post and should gain more traffic

File: flatbuffers/flatbuffers_case_study.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+table Event {
+  timestamp: ulong;
+  type: EventType;
+  event_data: EventData;
+}
+
+root_type Event;
+```
+
+After defining the schema, we run the flatbuffer 'compiler' - flatc. This generates code in our language of choice. That code handles all the bytes and bits calculation flatbuffer is based on.
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md]: one thing to mention is that flatbuffer are available to a LOT of languages aside from js

File: flatbuffers/flatbuffers_case_study.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+* A recorder
+* Processing pipeline
+* Recordings storage
+* Recordings player
+
+The recorder have the ability to capture any event or DOM mutation in a page, and stream the recorded entities to our servers. As you can imagine a session might consist of a huge number of events. Every mouse move, every scroll, every change to the page itself must be recorded. This can easily mount to ten or hundreds of thousands of data points.
+
+All this data must be stored and accessible for the recorders counterpart - the player. The player knows how to take the mass of events and reconstruct them into a watchable clip.
+
+This undertaking raises many challenges, one of which is to transfer all of those events from the recorded user to our servers, and from our servers to the player. The efficiency of this process is measured both by the size of the transfered data and by the time it takes to parse it, measures that JSON does not excel in.
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md:25]: I would also add a background check we did of FB vs other similar frameworks we did and why we chose flatbuffers..

File: flatbuffers/flatbuffers_case_study.md [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+const buffer = new flatbuffers.ByteBuffer(bytes)
+const event = EventSerializer.Event.getRootAsEvent(buffer)
+
+console.log('event timestamp', event.timestamp())
+console.log('event type', event.type())
+console.log('event data type', event.dataType())
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md]: maybe talk about actual cons? like no typesafe, building is form the outside in

File: flatbuffers/flatbuffers_case_study.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+The results where quite impressive! 40% - 94% size decrease in our test data.
+
+# Drawbacks
+The advantages FlatBuffers offers do not come for free.
+
+* Non human readable
+* Verbal and cumbersome API
+* Backwards compatibility issues
+* Lack of streaming support
```
Comment (BlueHotDog) [inline flatbuffers/flatbuffers_case_study.md]: I would write a conclusion
And also touch put in the benchmarks speed comparison.

**Review by BlueHotDog — COMMENTED** [review]
very awesome!

File: flatbuffers/flatbuffers_case_study.md (hunk 1 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: 👍

File: flatbuffers/flatbuffers_case_study.md (hunk 2 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: 👍

File: flatbuffers/flatbuffers_case_study.md (hunk 6 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md:25]: did we do that? what were the conclusions?

File: flatbuffers/flatbuffers_case_study.md (hunk 5 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: I've added it to the introduction

File: flatbuffers/flatbuffers_case_study.md (hunk 7 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: good point. I have a drawbacks section at the bottom, I'll add these points

File: flatbuffers/flatbuffers_case_study.md (hunk 8 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: 👍

File: flatbuffers/flatbuffers_case_study.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+console.log('event timestamp', event.timestamp())
+console.log('event type', event.type())
+console.log('event data type', event.dataType())
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
+
+# Phases of refactor
+Refactoring our entire flow from JSON to FlatBuffers is quite a feat. It's a architecture crossing change that will be dangerous to try and accomplish in one go. We've considered this issue for a while and managed to split the redactor into several independent steps.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: redactor => refactor

File: flatbuffers/flatbuffers_case_study.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+# What are FlatBuffers?
+[FlatBuffers](https://google.github.io/flatbuffers) is a serialization library by Google. It was first made by Google with android games in mind, but now it has ports to C++, C#, C, Go, Java, PHP, Python, and JavaScript.
+
+The main goal of FlatBuffers is to avoid deseriazation what so ever. This is achieved by defining a binary data protocol - a well defined method to convert data to, and from binary. The binary structure created by this protocol can be sent over the wire, and read without further processing. To reiterate, when transferring JSON, we need to transform data into a string, send it over the wire, parse the string, and transform it into a native object. Flatbuffers does not require any of it. You start up with binary, send the same binary, and read directly from binary.
+
+In addition to the parsing efficiency the binary format gains us another advantage - binary representation of data is often more size efficient. We can store a 10 digit int using 4 bytes UInt, instead of 10 chars.
+
+## Flat buffers Example
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: Flat buffers => FlatBuffers
I think this should be consistent in the entire article

File: flatbuffers/flatbuffers_case_study.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+This undertaking raises many challenges, one of which is to transfer all of those events from the recorded user to our servers, and from our servers to the player. The efficiency of this process is measured both by the size of the transfered data and by the time it takes to parse it, measures that JSON does not excel in.
+
+# What are FlatBuffers?
+[FlatBuffers](https://google.github.io/flatbuffers) is a serialization library by Google. It was first made by Google with android games in mind, but now it has ports to C++, C#, C, Go, Java, PHP, Python, and JavaScript.
+
+The main goal of FlatBuffers is to avoid deseriazation what so ever. This is achieved by defining a binary data protocol - a well defined method to convert data to, and from binary. The binary structure created by this protocol can be sent over the wire, and read without further processing. To reiterate, when transferring JSON, we need to transform data into a string, send it over the wire, parse the string, and transform it into a native object. Flatbuffers does not require any of it. You start up with binary, send the same binary, and read directly from binary.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: deseriazation => deserialization

File: flatbuffers/flatbuffers_case_study.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+In addition to the parsing efficiency the binary format gains us another advantage - binary representation of data is often more size efficient. We can store a 10 digit int using 4 bytes UInt, instead of 10 chars.
+
+## Flat buffers Example
+This won't be a FlatBuffers tutorial, but to give you a feel for what it's like lets took on a simplified version of our data model.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: took on => look at

File: flatbuffers/flatbuffers_case_study.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+table Event {
+  timestamp: ulong;
+  type: EventType;
+  event_data: EventData;
+}
+
+root_type Event;
+```
+
+After defining the schema, we run the flatbuffer 'compiler' - flatc. This generates code in our language of choice. That code handles all the bytes and bits calculation flatbuffer is based on.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: flatbuffer => FlatBuffers

File: flatbuffers/flatbuffers_case_study.md [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+const buffer = new flatbuffers.ByteBuffer(bytes)
+const event = EventSerializer.Event.getRootAsEvent(buffer)
+
+console.log('event timestamp', event.timestamp())
+console.log('event type', event.type())
+console.log('event data type', event.dataType())
+console.log('event data', event.event_data(new EventSerializer.MousePositionData())).x()
+```
+
+As you can see, Flatbuffers creation and reading is not trivial. The API is verbal and cumbersome.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: Flatbuffers => FlatBuffers

File: flatbuffers/flatbuffers_case_study.md [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+# Phases of refactor
+Refactoring our entire flow from JSON to FlatBuffers is quite a feat. It's a architecture crossing change that will be dangerous to try and accomplish in one go. We've considered this issue for a while and managed to split the redactor into several independent steps.
+
+## FlatBuffers in recorder
+We began in transforming our recorder messages to FlatBuffers. The recorder creates and send the recorded events as FlatBuffers. The first step of the processing pipeline is to transform the binary data back to JSON allowing the rest of the architecture to stay unchanged.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: send => seds

File: flatbuffers/flatbuffers_case_study.md [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+## FlatBuffers in player
+The second step we took was to add FlatBuffers support to our player. Naturally we needed to convert the JSON data we store to FlatBuffers but this extra converting code is a small price to pay in order to factorize a big refactor.
+
+## Store FlatBuffers - "point of no return"
+The Final step is to actually store the FlatBuffers bytes in our storage. This is the "point of no return". In previous steps we could easily roll back the change. Once we loose JSON all together, a rollback would be very difficult.
+
+In fact we have not taken this step yet. A Major benefit of JSON is that it's easily streamable using libraries such as [Oboe.js](http://oboejs.com/). We have yet to discover a method to achieve a similar goal using FlatBuffers.
+
+# benchmarks
+## Space banchmark
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: Space banchmark => Storage size benchmark

File: flatbuffers/flatbuffers_case_study.md [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+We have expected FlatBuffers to reduce the size of our data. We've converted some of our JSON sessions to FlatBuffers to check this hypothesis. Each tested session is from a different customer, and has different characteristics, so variation in the results are expected.
+
+<scatter plot>
+
+The result did not met our expectations, not only the FlatBuffers file are not smaller that our JSON, they are 15%-30% bigger. While this is a disappointing result the overall file size is just a secondary goal. This issue will definitely be studies.
+
+## Data availability speed benchmark
+The next important benchmark would be data availability speed. By this I mean the duration of time it takes from initial request to complete data availability. In this benchmark we will take several session, in both JSON and FlatBuffers format, download and iterate over their data to perform some simple calculation.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: several session => several sessions

File: flatbuffers/flatbuffers_case_study.md [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+
+<some kind of graph>
+
+As expected, JSON is impeded by the parsing stage. FlatBuffers allows immediate access to data after download. Those several seconds of JSON parsing are critical for us. These are the first seconds of our page loads, where the first impression of our system takes place.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: `These are the first seconds of our page loads, where the first impression of our system takes place.`
something is off with this sentence

File: flatbuffers/flatbuffers_case_study.md [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+* Backwards compatibility issues - As we deal with structured binary data we must consider the possibility of a change to that structure. Adding or removing fields from our schema must be done carefully. A wrong schema change might cause silent errors while reading old version objects. [Some good pointers about how to do this right exist in the FlatBuffers Docs](https://google.github.io/flatbuffers/flatbuffers_guide_writing_schema.html)
+
+* Lack of streaming support - When dealing with large amount of data there is a good chance you would like to stream it. There are several issues with streaming a FlatBuffer array. Flatbuffers are written backwards. This means that crucial bits of our data appear at the end of the file, making streaming not feasible. Even if this issue could be overcome another hurdle exists. XmlHttpRequest does not support streaming of binary data, meaning that we'll have to use the new Fetch API. The new Fetch API is awesome, but not supported in all major browsers.
+
+# Conclusion
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md:166]: we pretty pleased => we are pretty pleased

File: flatbuffers/flatbuffers_case_study.md [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+Several months into the FlatBuffers refactor we pretty pleased with the result. We've seen a noticeable performance improvement in the areas of the architecture we've converted to FlatBuffers.
+
+This was not an easy ride. There are several caveats in FlatBuffers that gets exaggerated by JavaScatips weak typing nature.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: JavaScatips => Javacript's

File: flatbuffers/flatbuffers_case_study.md [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+FlatBuffer is heavy machinery. It won't do you any good if you don't have a vast number of objects to throw at it.
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md]: FlatBuffer => FlatBuffers ?

**Review by itayadler — CHANGES_REQUESTED** [review]

File: flatbuffers/flatbuffers_case_study.md (hunk 4 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: You're right, but the post talks about our specific experience. What about "Refactoring JSON to FlatBuffers", "Moving from JSON to FlatBuffers", "Case study of JSON to FlatBuffers transition". Any ideas?

Comment (itayadler) [reply flatbuffers/flatbuffers_case_study.md]: I like Refactoring JSON to FlatBuffers

File: flatbuffers/flatbuffers_case_study.md (hunk 11 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 10 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 12 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 13 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 14 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 9 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 15 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 16 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 17 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 18 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: how do you think it should be?

File: flatbuffers/flatbuffers_case_study.md (hunk 19 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md:166]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 20 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 21 above)
Comment (alexpusch) [reply flatbuffers/flatbuffers_case_study.md]: done

File: flatbuffers/flatbuffers_case_study.md (hunk 15 above)
Comment (itayadler) [reply flatbuffers/flatbuffers_case_study.md]: bah seds :) I mean't sends

File: flatbuffers/flatbuffers_case_study.md (hunk 18 above)
Comment (itayadler) [reply flatbuffers/flatbuffers_case_study.md]: Those several seconds of JSON parsing are critical for us. These are the first seconds of our page loads, where the first impression of our system takes place. 
=>
The time spent on JSON parsing during page load time is noticeable when dealing with a higher volume of data, and if we can shave it off, then that'll be key to delivering a non-janky experience to our users.

What do you think?

File: flatbuffers/flatbuffers_case_study.md [hunk 22]
```diff
@@ -138,19 +138,19 @@ The Final step is to actually store the FlatBuffers bytes in our storage. This i
 In fact we have not taken this step yet. A Major benefit of JSON is that it's easily streamable using libraries such as [Oboe.js](http://oboejs.com/). We have yet to discover a method to achieve a similar goal using FlatBuffers.
 
 # benchmarks
-## Space banchmark
+## storage benchmark
```
Comment (itayadler) [inline flatbuffers/flatbuffers_case_study.md:141]: Captial S

**Review by itayadler — CHANGES_REQUESTED** [review]

**Pull request merged**